smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
+            Points at which the model is evaluated
 
         Returns
         -------
-        list (for dimension > 1) or float (for dimension = 1)
-            Derivative of the surrogate in all dimensions.
+        ndarray of shape (n_samples,) or (n_samples, n_targets)
+            Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
+        ValueError
             Input must lie in domain of surrogate.
-        IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
-        else:
-            oob = (numpy.any(x < self.domain[0]) or 
-                    numpy.any(x > self.domain[1]))
-        if oob:
-            raise ValueError('x must lie in domain of surrogate')
-        # transform point into basis domain
-        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
-        # evaluate the gradient
-        gradient = numpy.zeros(x_scaled.shape)
-        for ni in range(self.dimension):
-            value = 0
-            for coeff, basis in zip(self.coefficients,
-                                    self.grid.basis_functions):
-                if self.dimension > 1:
-                    term = numpy.prod(
-                            [basis[dim].derivative(x_scaled[...,dim]) 
-                             if dim == ni else basis[dim](x_scaled[...,dim]) 
-                             for dim in range(len(basis))], axis=0)
-                else:
-                    term = basis.derivative(x_scaled)
-                
-                value += coeff*term
-            gradient[...,ni] = numpy.reshape(value, gradient.shape[:-1])
 
-        if len(input_shape) == 0:
-            gradient = gradient.item()
-        else:
-            if (self.dimension == 1 and len(input_shape) > 1 and input_shape[-1] == 1):
-                output_shape = input_shape[:-1]
-            else:
-                output_shape = input_shape
-            gradient = numpy.reshape(gradient, output_shape)
-        return gradient
 
-    def __call__(self, x):
+class ProductSetSurrogate(BaseSurrogate):
```

Comment:
I think this a set product, not a product set, right?
```suggestion
class SetProductSurrogate(Surrogate):
```